pkg/apis/storage/types.go

```diff
@@ -672,3 +672,50 @@ type CSIStorageCapacityList struct {
 	// Items is the list of CSIStorageCapacity objects.
 	Items []CSIStorageCapacity
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VolumeAttributesClass represents a specification of mutable volume attributes
+// defined by the storage driver. The class can be specified during dynamic provisioning
+// of PersistentVolumeClaims, and changed after provisioning.
+type VolumeAttributesClass struct {
+	metav1.TypeMeta
+
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// Name of the CSI driver
+	DriverName string
+
+	// parameters hold volume attributes defined by the CSI driver. These values
+	// are opaque to the system and are passed directly to the CSI driver.
+	// The underlying storage provider supports changing these attributes on an
+	// existing volume, however the parameters field itself is immutable. To
+	// invoke a volume update, a new VolumeAttributesClass should be created with
+	// new parameters, and the PersistentVolumeClaim should be updated to reference
+	// the new VolumeAttributesClass.
+	//
+	// The keys cannot be empty, and the maximum number of parameters is 512, with
+	// a cumulative max size of 256K.
```

Comment:
Are keys not defined by the CSI driver allowed? What happens when there is a key in the map that the CSI driver does not recognize? (Use cases I'm interested in: (a) user accidentally has a typo in a map key string (b) user adds a map key that a newer version of a CSI driver supports, but the CSI driver in use is an older version).

Reply:
It depends on the implemention of a special CSI driver. When the driver recieves unrecognize params, it can ingore those unwant params or reject the request with the INVALID_ARGUMENT error code.

FYI:

- CreateVolume Errors: https://github.com/container-storage-interface/spec/blob/master/spec.md#createvolume-errors
- ControllerModifyVolume Errors: https://github.com/container-storage-interface/spec/blob/master/spec.md#controllermodifyvolume-errors

Reply:
I think we can say if the CSI driver rejects invalid parameters, then we sill set the pvc.modifyvolumestatus = Infeasible